<a href="https://colab.research.google.com/github/MattWroclaw/data-science-bootcamp/blob/main/Uczenie_maszynowe/scikit_learn_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import sklearn
sklearn.__version__

'1.3.2'

### Klasyfikacja

In [2]:
import numpy as np
from sklearn import datasets
# from sklearn.model_selection import train_test_

np.random.seed(10)
raw_data = datasets.load_iris()
raw_data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [3]:
raw_data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [4]:
print(raw_data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [7]:
data = raw_data.data
target = raw_data.target

In [8]:
print(data.shape)
print(target.shape)

(150, 4)
(150,)


(150, 4)
(150,)  
150 próbek  
4 zmienne  

### Należy podzielić dane na zbiór *treningowy* i *testowy*

In [9]:
from sklearn.model_selection import train_test_split

#  ta funcja zwraca 4 obiekty
data_train, data_test, target_train, target_test = train_test_split(data, target, test_size=0.3)

In [10]:
print(data_train.shape)
print(data_test.shape)
print(target_train.shape)
print(target_test.shape)

(105, 4)
(45, 4)
(105,)
(45,)


In [11]:
data_test

array([[6.3, 2.3, 4.4, 1.3],
       [6.4, 2.7, 5.3, 1.9],
       [5.4, 3.7, 1.5, 0.2],
       [6.1, 3. , 4.6, 1.4],
       [5. , 3.3, 1.4, 0.2],
       [5. , 2. , 3.5, 1. ],
       [6.3, 2.5, 4.9, 1.5],
       [5.8, 2.7, 4.1, 1. ],
       [5.1, 3.4, 1.5, 0.2],
       [5.7, 2.8, 4.5, 1.3],
       [5.6, 3. , 4.5, 1.5],
       [5.8, 2.7, 5.1, 1.9],
       [5.5, 2.3, 4. , 1.3],
       [4.9, 3. , 1.4, 0.2],
       [5.1, 3.8, 1.5, 0.3],
       [6.8, 3. , 5.5, 2.1],
       [6. , 3.4, 4.5, 1.6],
       [4.4, 3. , 1.3, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [5. , 3.2, 1.2, 0.2],
       [7.1, 3. , 5.9, 2.1],
       [6.4, 2.8, 5.6, 2.2],
       [6.2, 2.8, 4.8, 1.8],
       [4.8, 3.4, 1.9, 0.2],
       [5.9, 3. , 4.2, 1.5],
       [4.7, 3.2, 1.3, 0.2],
       [5.7, 3. , 4.2, 1.2],
       [5.5, 2.6, 4.4, 1.2],
       [6.8, 2.8, 4.8, 1.4],
       [7.7, 3.8, 6.7, 2.2],
       [6.6, 2.9, 4.6, 1.3],
       [6.2, 2.9, 4.3, 1.3],
       [7.2, 3. , 5.8, 1.6],
       [5.8, 2.8, 5.1, 2.4],
       [6.3, 2

In [12]:
# import klasy która przechowuje model ; LogisticRegression - model klasyfikacji
from sklearn.linear_model import LogisticRegression

# tworzymy instancje LogisticRegression
model = LogisticRegression()

# metoda do trenowania modelu
model.fit(data_train, target_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [14]:
# predict przewidzenie wartości zbioru testowego; przekazujemy tylko cechy targetu, nie przekazujemy co to za target

target_pred = model.predict(data_test)
target_pred

array([1, 2, 0, 1, 0, 1, 1, 1, 0, 1, 1, 2, 1, 0, 0, 2, 1, 0, 0, 0, 2, 2,
       2, 0, 1, 0, 1, 1, 1, 2, 1, 1, 2, 2, 2, 0, 2, 2, 2, 2, 0, 0, 1, 0,
       1])

In [15]:
# użyjemy metric do oceny modelu; importujemy macież konfuzji. Przyjmuje ona target_test oraz y_pred

from sklearn.metrics import confusion_matrix

confusion_matrix(target_test, target_pred)

array([[14,  0,  0],
       [ 0, 17,  0],
       [ 0,  0, 14]])

Macierz konfuzji powie nam ile błędów popełnił model na danych testowych  
Klasę pierwszą przewidział  w 100% pewnie [14,0,0], w klasie drugiej miał 5 błędów [0,12,5]

**Interesuje nas maksymalizacja wartości na przekątnej macierzy**

In [16]:
# kolejna ocena stopnia nauczenia się modelu
from sklearn.metrics import classification_report

print(classification_report(target_test, target_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        14
           1       1.00      1.00      1.00        17
           2       1.00      1.00      1.00        14

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45



# Uwaga
**Najczęściej spędza się bardzo wiele czasu do przygotowania danych** zanim zaczniemy trenować model. I trzeba podzielić zbiór na 2 grupy


1.   Treningowe dane
2.   Testowe dane - na nich oceniamy model

Model nie może widzieć danych testowych.

# Wzrór postępowania


1.   Tworzymy obiekt klasy z której będziemy korzystać

```
model = LogisticRegression()
```


2.   Trenowanie modelu  


```
model.fit()
```
Do fit przekazujemy dane treningowe
```
# data_train -> cechy, dane niezależne
# target_train -> do tego przyporządkowujemy te cechy
fit(data_train, target_train)

```

3.   Przewidzianie danych

```
model.predict()
```
do predict przekazujemy zmienne testowe
```
model.predict(data_test)
```

Dzięki temu dostajemy etykiety ze zbioru testowego.

4.   Ocena czy to co przewidział model zgadza się z faktycznym opisem
Możemy to zrobić za pomocą `macierzy konfuzji` albo `classification_report`



**Uwaga** Modele klasyfikacji oraz modele regresji mają swoje specyficzne sposoby oceny